astropy/io/fits/hdu/hdulist.py

```diff
@@ -912,11 +914,9 @@ def writeto(self, fileobj, output_verify='exception', overwrite=False,
         # of the caller)
         closed = isinstance(fileobj, str) or fileobj_closed(fileobj)
 
-        # writeto is only for writing a new file from scratch, so the most
-        # sensible mode to require is 'ostream'.  This can accept an open
-        # file object that's open to write only, or in append/update modes
-        # but only if the file doesn't exist.
-        fileobj = _File(fileobj, mode='ostream', overwrite=overwrite)
+        # This can accept an open file object that's open to write only, or in
+        # append/update modes but only if the file doesn't exist.
+        fileobj = _File(fileobj, mode=mode, overwrite=overwrite)
```

Comment:
@drdavella - Do we really need a `mode` keyword in addition to the fileobj's mode ? I don't think that we want this, as it would be really confusing. And the goal of the FITS mode / file mode mapping is to avoid this. 
The deleted comment above says that ostream should work also for "append/update modes but only if the file doesn't exist", we could maybe find why it is not the case ?

Reply:
The issue comes from the check `objmode = _normalize_fits_mode(fileobj_mode(fileobj))`, which brings us back to https://github.com/astropy/astropy/commit/973f94ba0955f212744aac98e5504a5b992b797a ;)
We could maybe just add a special case for this ?

Reply:
> Do we really need a mode keyword in addition to the fileobj's mode ? I don't think that we want this, as it would be really confusing.

This is the situation that we have already with `_File` and `fits.open`, and we're stuck with those. Unfortunately, I think that adding `mode` is the right course of action here, because the previous behavior of assuming `mode='ostream'` for all uses of `writeto` is incorrect. Luckily, most users never have to think about this since the default of `ostream` seems reasonable for most cases. But #7854 demonstrates that the assumption is not correct for **all** cases.

As for `_normalize_fits_mode` and the commit/PR you referenced: the point of those changes was that the mode checks were previously not being performed correctly in all cases. Remember, we already let users pass `mode=` to `_File` and `fits.open` even though they can pass a file object with its own mode. So we have to perform the compatibility check. I believe we are now performing mode checks correctly. That PR exposed some incompatible uses in the FITS code base. #7854 is another case that was missed because there wasn't a test case. If we add some sort of special case here, I'm not sure we can guarantee that we will continue to check the mode compatibility correctly in all cases.

However, I'm open to suggestions if you have a particular idea of what the special case would look like.

Reply:
Btw, I agree that this is quite confusing, both for users and developers. I'm not quite sure why FITS needs to make reference to modes like `ostream`, `append`, at all. Why can't we just be consistent with the Python API? There's probably a good reason, but in any case changing this would require a complete overhaul of the `io.fits` API, which is probably a non-starter 😉.

Reply:
Keep in mind that `astropy.io.fits` was PyFITS, and PyFITS existed way way back when Python couldn't do some of the stuff it can do now. 😉

Reply:
> because the previous behavior of assuming mode='ostream' for all uses of writeto is incorrect

But given the comment above, I think that this was working before the addition of `_normalize_fits_mode`. I know that this check was added for good reasons, and we need to check the modes correctly, but it is just a check and nothing prevent to allow `ostream` to work with `rb+` or `ab+`.

Currently we have this `mode` keyword only for `fits.open` (`_File` is private), and I think the reason is just to provide more user friendly names. But for all the other methods, from an API point of view, it would be weird, and annoying to have to provide `mode` if you already opened the file with a given file mode. 
So I think that we could either guess the FITS mode from the file at each place where it is necessary (what you did for `fits.append` above), or allow `ostream` to work with `rb+` or `ab+`.

Reply:
It works with 2.0 (before `_normalize_fits_mode`):
```
In [1]: %astropy
Numpy 1.15.2
Astropy 2.0.8

In [2]: with open('foo.fits', mode='ab+') as f:
   ...:     fits.append(f, data=np.zeros(4))
   
In [3]: fits.info('foo.fits')
Filename: foo.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU       5   (4,)   float64   
```
So doing a less stringent test should work just as before without adding new bugs or side effects imo. If you prefer adding the mode guessing as in `fits.append`, I'm also fine with it, but I think that we should avoid adding a new keyword, even if it is hidden.

Reply:
It's not at all clear to me how to handle this particular case without performing a less stringent test in all cases. I would argue that downgrading this check means that we don't actually care whether we conform to the API as described. The consequence will be that the behavior of the `mode` argument to `fits.open` is sometimes unpredictable.

Keep in mind that the changes associated with `_normalize_fits_mode` were not unmotivated: the API was actually incorrect prior to that change. So we're saying we're okay with reverting to that state, and we will need to remove and/or modify some associated unit tests.

Reply:
BTW, I realize I must seem quite stubborn to you, but I think it boils down to this: you appear to resist introducing complexity into the API, which I definitely respect, but I believe in this case it would mean sacrificing correctness and consistency in other parts of the API.

Reply:
@drdavella - I know this was added for good reasons, and I understand your concerns - except for the emphasis about the API. For me it is mostly an internal thing, that users should not care about. My concern is that in terms of "public" API, something like this was working (in 2.0):
```python
with open('foo.fits', mode='ab+') as f:
    hdu = fits.ImageHDU(data=np.ones(5))
    hdu.writeto(f)
```
and now fails, and with the change that you propose here, users would have to use this:
```python
with open('foo.fits', mode='ab+') as f:
    hdu = fits.ImageHDU(data=np.ones(5))
    hdu.writeto(f, mode="append")
```
or this
```python
with open('foo.fits', mode='wb+') as f:
    hdu = fits.ImageHDU(data=np.ones(5))
    hdu.writeto(f, mode="update")
```
This is a weird API for me ;). So I think that we can find a way to manage this internally, even if this needs some sacrificen, it will not be the first hack in the io.fits code. (And I would be happy if we had a better solution but I don't see one currently...)

Reply:
Okay yes, I agree that those cases are definitely not desirable. My hope/intention was that the `mode` argument to `*HDU.writeto` would be completely unnecessary and invisible to users for most or all use cases, but that's apparently not the case.

I'll add these examples as test cases (always good to have more of those!) and will try to work out a better solution here.